cms/server/contest/handlers/taskusertest.py

```diff
@@ -302,15 +304,29 @@ def post(self, task_name):
         # If we allow partial submissions, implicitly we recover the
         # non-submitted files from the previous submission. And put them
         # in file_digests (i.e. like they have already been sent to FS).
-        submission_lang = None
+        submission_lang = self.get_argument("submission_lang", "auto")
+        chosen_language = submission_lang
         file_digests = {}
-        if task_type.ALLOW_PARTIAL_SUBMISSION and last_user_test_t is not None:
+        if task_type.ALLOW_PARTIAL_SUBMISSION and \
+                last_user_test_t is not None:
             for filename in required.difference(provided):
                 if filename in last_user_test_t.files:
                     # If we retrieve a language-dependent file from
                     # last submission, we take not that language must
                     # be the same.
                     if "%l" in filename:
+                        if chosen_language != self.AUTO \
+                            and last_user_test_t.language is not None and \
+                                chosen_language != last_user_test_t.language:
+                            self.application.service.add_notification(
+                                participation.user.username,
+                                self.timestamp,
+                                self._("Invalid submission!"),
+                                self._("All sources must be"
+                                       " in the same language."),
+                                NOTIFICATION_ERROR)
+                            self.redirect("/testing?%s"
+                                          % quote(task.name, safe=''))
```

Comment:
This seems so wrong! Why can't a user send tests using different languages? It's not like we use it for scoring or etc. I checked and it seems that the problem exists prior to your changes as well. I think it'd be better if we'd fix it once and for all :)